Fixes #27877 - no conflicts on DHCP filename change

lzap opened the pull request:
Our orchestration code checks existing DHCP records against MAC, IP, hostname and filename. However filename must not be included when comparing DHCP records for conflicts - it is totally possible to create multiple records with the same filename option. It should have never been considered as a subject of conflict. This change is about removing the filename from the check so it won't issue unnecessary conflict errors on updates of PXE loader.

lzap committed:
Fixes #27877 - no conflicts on DHCP filename change

ekohl commented on lib/net/dhcp/record.rb:
```diff
@@ -75,9 +76,6 @@ def ==(other)
         to_compare << :hostname if other.attrs[:hostname].present? && attrs[:hostname].present?
       end
 
-      # Not all DHCP smart-proxy providers support TFTP filename option (e.g. libvirt).
```
Might still be nice to compare these to know if there would be a change. I don't know if we do a `old.record != record && update_record(record)`. Probably better to not rely on this method at all in `conflicts`.

lzap replied:
I am not following, it is incorrect to include filename in comparsion as it cannot generate conflicts in DHCP servers. The comment was just statement about filename option which is not very relevant to the problem.

ekohl replied:
This method tests for equality. This means that the filename is no longer considered in testing whether 2 records are equal. If you have code that checks if an update is needed by doing `if old.record != record ; update_record(record) ; end` then this change means you will miss updates. That's why I was suggesting that the `conflicts` method shouldn't rely on equality comparison.

ezr-ondrej replied:
I haven't noticed it in the codebase, but maybe I missed that and make sense.

lzap commented:
@timogoebel would you mind taking a look?

timogoebel reviewed (dismissed):
Makes sense to me. I haven't tested this, though.

ezr-ondrej reviewed (approved):
Make sense, works well. Could you please rebase?

ezr-ondrej commented on lib/net/dhcp/record.rb:
```diff
@@ -52,10 +52,11 @@ def create
     def conflicts
       conflicts = [proxy.record(network, mac), proxy.records_by_ip(network, ip)].flatten.compact.delete_if { |c| c.lease? || c == self || related_macs.include?(c.mac) }
-      @conflicts ||= conflicts.uniq {|c| c.attrs}
+      @conflicts ||= conflicts.uniq
```
@ekohl has a point we do not need to rely on `==`. Could we keep that as is and do this instead?
```suggestion
      @conflicts ||= conflicts.uniq { |c| c.attrs.slice(:mac, :ip, :network) }
```

ekohl replied:
@ezr-ondrej that doesn't work right, because the line above selects records and the `c == self` line does the comparison. That means you'll still end up with the same the entry while `@conflicts` should be empty when there are none.

ezr-ondrej replied:
Ah, I see... so we would need to implement a method like this:
```ruby
def conflicts?(other)
  attrs.slice(:mac, :ip, :network) == other.attrs.slice(:mac, :ip, :network)
end
```
and use it, right? But I believe than the `==` operator would be useless :thinking:

lzap replied:
I dont understand, this completely drops the current behavior of skipping leases and related MACs.

ezr-ondrej reviewed (changes requested):
Ok @lzap what do you think about this?

ezr-ondrej commented:
@lzap what is the status here? :)

lzap commented:
I think the concern of overriding `==` is valid, I've created an explicit method and call it. Other than that this is ready for testing.

lzap commented on lib/net/dhcp/record.rb:
```diff
@@ -52,10 +52,13 @@ def create
     def conflicts
-      conflicts = [proxy.record(network, mac), proxy.records_by_ip(network, ip)].flatten.compact.delete_if { |c| c.lease? || c == self || related_macs.include?(c.mac) }
-      @conflicts ||= conflicts.uniq {|c| c.attrs}
+      conflicts = [proxy.record(network, mac), proxy.records_by_ip(network, ip)].flatten.compact.delete_if { |c|
```
Dumb rule.

lzap reviewed (commented):
Rebased.

ekohl commented on lib/net/dhcp/record.rb:
```diff
@@ -64,7 +67,7 @@ def valid?
     end
 
-    def ==(other)
+    def eql_for_conflicts?(other)
```
Should we keep the old method? Does something rely on it?

ezr-ondrej replied:
It doesn't seems so, but all the tests are testing the old method, we should change the tests to test the new one instead.

ezr-ondrej commented:
@lzap the tests are testing the old method, so now they are failing. We should change the tests to use the new method instead.

lzap commented:
Fixed tests.

ekohl commented on lib/net/dhcp/record.rb:
```diff
@@ -52,19 +52,22 @@ def create
     def valid?
       logger.info "Fetching DHCP reservation #{name} for #{self}"
-      self == proxy.record(network, mac)
+      self.eql_for_conflicts?(proxy.record(network, mac))
```
I think you may be breaking the API here. This called at:
https://github.com/theforeman/foreman/blob/4485b8e8f92bafc5eb7d3ae5858985d4a8c9079c/app/models/concerns/orchestration/dhcp.rb#L187-L189
Then that's called by
https://github.com/theforeman/foreman/blob/4485b8e8f92bafc5eb7d3ae5858985d4a8c9079c/app/models/concerns/orchestration/dhcp.rb#L171-L185

By making this change, the DHCP orchestration will no longer detect that the filename needs to be updated.

IMHO the `==` method was still the correct one and removing it here is not the right thing to do. Am I missing something?

lzap replied:
It works, I don't understand.

Edit: Note in dhcp.rb we compare individual fields but not the object, therefore this code does not call the old overloaded method.

ekohl replied:
The point I tried to make is that there's 2 flows:
* Is there a conflicting DHCP record? (what you're implementing here)
* Does the existing DHCP record need an update? (what I'm linking in my above comment)

My point is about the latter flow. To spell it out the call graph:
* `dhcp_update_required?` calls `all_dhcp_records_valid?`
* `all_dhcp_records_valid?` checks whether there's any mismatch. Your code change means it ignores the filename

This means you are breaking the API that `all_dhcp_records_valid?` expects. This may not be an issue because there's also a `operatingsystem.boot_filename(old.host) != operatingsystem.boot_filename(self.host)` check.

What I'm suggesting that the API is much cleaner if you have two methods on the DHCP record:
* `conflicts` checks whether there's a conflicting record
* `==(other)` checks if the record is exactly the same. If it's not, it needs an update.

That would make the whole orchestration much easier to follow.

ekohl reviewed (commented):
Looking a bit further, this has a `legacy_dhcp_api?` method which makes me wonder if we can get rid of that. It looks like this was introduced in smart-proxy 1.16.0 so that's probably safe to rely on by now. Candidate for another PR?

adamruzicka commented:
What is the status here?

ekohl commented:
I should have updated the labels, done so now.

lzap commented:
Sorry, I don't understand what you mean, you wanted me to get rid of `==` method, now you say I should be putting it back. Please elaborate.

This patch has been working with or without `==` (the current version) and I believe is ready to be merged. When PXE loader is changed, DHCP orchestration correctly schedules update. This is all about conflict checks and the description is I believe pretty clear.

```
host sarah-killius.nat.lan {
  dynamic;
  hardware ethernet 52:54:00:13:13:19;
  fixed-address 192.168.99.158;
        supersede server.filename = "grub2/grubx64.efi";
        supersede server.next-server = c0:a8:63:01;
        supersede host-name = "sarah-killius.nat.lan";
}
host sarah-killius.nat.lan {
  dynamic;
  deleted;
}
host sarah-killius.nat.lan {
  dynamic;
  hardware ethernet 52:54:00:13:13:19;
  fixed-address 192.168.99.158;
        supersede server.filename = "pxelinux.0";
        supersede server.next-server = c0:a8:63:01;
        supersede host-name = "sarah-killius.nat.lan";
}
```

ezr-ondrej reviewed (dismissed):
@ekohl I stand with @lzap here, I don't see where the equality operator is being used elsewhere and I like this implementation.

lzap commented:
@ezr-ondrej can you merge please? I need to get this one moving.

adamruzicka commented:
@lzap There is a conflicting file, could you rebase please?

tbrisker commented:
please rebase @lzap

lzap commented:
Rebased.

lzap commented:
Tests are green.

adamruzicka commented:
Is this waiting for something?

ezr-ondrej commented:
> Is this waiting for something?

I need to test if the @ekohl's concern (https://github.com/theforeman/foreman/pull/7047#discussion_r385599337) is valid.
Any help appreciated.

ekohl commented:
>  Any help appreciated.

What kind of help are you looking for?

ezr-ondrej commented:
> What kind of help are you looking for?

@ekohl What @lzap says and what you say contradicts: 
https://github.com/theforeman/foreman/pull/7047#issuecomment-589682745 says it works for both - conflict detection and update planning, but you - https://github.com/theforeman/foreman/pull/7047#discussion_r385599337 say it should not work now.
I need to test and dig deeper

ekohl commented:
I think in most cases it'll be covered by `operatingsystem.boot_filename(old.host) != operatingsystem.boot_filename(self.host)`. However, if a boot filename was written in Foreman x.y and the logic changes in Foreman x.z then I suspect it'll not detect `operatingsystem.boot_filename(self.host) != $actual_boot_filename_in_dhcp`.

ezr-ondrej reviewed (approved):
The orchestration does indeed deserve some attention and cleanup, as there is a point, we don't detect the bootfile content change. This patch shouldn't make it worse though.
Thanks @lzap, merging!

lzap commented:
> we don't detect the bootfile content change

Yes, I have a BZ on my plate and I will be looking into that shortly. Thanks. https://bugzilla.redhat.com/show_bug.cgi?id=1811946
